File: tools/nia_tools/report/crate_boundaries.py
#!/usr/bin/env python3
"""Report evidence for reviewing Nia workspace crate boundaries."""

from __future__ import annotations

import argparse
import json
import re
import subprocess
from collections.abc import Sequence
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Iterable

from tools.nia_tools.repository import REPOSITORY_ROOT

ROOT = REPOSITORY_ROOT
PUBLIC_ITEM = re.compile(
    r"^pub\s+(?:async\s+|unsafe\s+|const\s+|extern\s+)*"
    r"(?:struct|enum|union|trait|type|const|static|fn|mod|use)\b"
)


@dataclass(frozen=True)
class CrateBoundary:
    name: str
    rust_loc: int
    public_items: int
    production_dependencies: tuple[str, ...]
    production_dependents: tuple[str, ...]
    dev_only_dependents: tuple[str, ...]


def cargo_metadata(root: Path) -> dict[str, Any]:
    result = subprocess.run(
        ["cargo", "metadata", "--no-deps", "--format-version", "1"],
        cwd=root,
        check=True,
        stdout=subprocess.PIPE,
        text=True,
    )
    return json.loads(result.stdout)


def dependency_kinds(dependency: dict[str, Any]) -> set[str]:
    kinds = dependency.get("kind")
    if kinds is not None:
        return {kinds}
    detailed_kinds = dependency.get("dep_kinds")
    if detailed_kinds:
        return {entry.get("kind") or "normal" for entry in detailed_kinds}
    return {"normal"}


def rust_source_metrics(crate_root: Path) -> tuple[int, int]:
    rust_loc = 0
    public_items = 0
    source_root = crate_root / "src"
    if not source_root.is_dir():
        return rust_loc, public_items
    for path in sorted(source_root.rglob("*.rs")):
        for line in path.read_text(encoding="utf-8").splitlines():
            if not line.strip():
                continue
            rust_loc += 1
            if PUBLIC_ITEM.match(line.strip()):
                public_items += 1
    return rust_loc, public_items


def workspace_boundaries(
    metadata: dict[str, Any],
) -> list[CrateBoundary]:
    packages = {
        package["name"]: package
        for package in metadata["packages"]
        if package["id"] in metadata["workspace_members"]
    }
    production_dependencies = {name: set() for name in packages}
    production_dependents = {name: set() for name in packages}
    dev_dependents = {name: set() for name in packages}

    for consumer, package in packages.items():
        for dependency in package["dependencies"]:
            provider = dependency["name"]
            if provider not in packages:
                continue
            kinds = dependency_kinds(dependency)
            if kinds - {"dev"}:
                production_dependencies[consumer].add(provider)
                production_dependents[provider].add(consumer)
            if "dev" in kinds:
                dev_dependents[provider].add(consumer)

    boundaries = []
    for name, package in packages.items():
        crate_root = Path(package["manifest_path"]).parent
        rust_loc, public_items = rust_source_metrics(crate_root)
        dev_only = dev_dependents[name] - production_dependents[name]
        boundaries.append(
            CrateBoundary(
                name=name,
                rust_loc=rust_loc,
                public_items=public_items,
                production_dependencies=tuple(sorted(production_dependencies[name])),
                production_dependents=tuple(sorted(production_dependents[name])),
                dev_only_dependents=tuple(sorted(dev_only)),
            )
        )
    return sorted(boundaries, key=lambda boundary: boundary.name)


def joined(values: Iterable[str]) -> str:
    return ",".join(values) or "-"


def write_tsv(boundaries: Iterable[CrateBoundary]) -> None:
    print(
        "crate\trust_loc\tpublic_items\tproduction_dependencies\t"
        "production_dependents\tdev_only_dependents"
    )
    for boundary in boundaries:
        print(
            f"{boundary.name}\t{boundary.rust_loc}\t{boundary.public_items}\t"
            f"{joined(boundary.production_dependencies)}\t"
            f"{joined(boundary.production_dependents)}\t"
            f"{joined(boundary.dev_only_dependents)}"
        )


def parse_args(arguments: Sequence[str] | None = None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        prog="python3 -m tools report crate-boundaries",
        description=(
            "Report deterministic crate size and workspace dependency evidence. "
            "Counts non-empty Rust lines and lexical public item declarations in src/."
        )
    )
    parser.add_argument(
        "--max-rust-loc",
        type=int,
        help="show only crates at or below this source size",
    )
    parser.add_argument(
        "--max-production-dependents",
        type=int,
        help="show only crates with at most this many production consumers",
    )
    return parser.parse_args(arguments)


def main(arguments: Sequence[str] | None = None) -> int:
    args = parse_args(arguments)
    boundaries = workspace_boundaries(cargo_metadata(ROOT))
    if args.max_rust_loc is not None:
        boundaries = [
            boundary
            for boundary in boundaries
            if boundary.rust_loc <= args.max_rust_loc
        ]
    if args.max_production_dependents is not None:
        boundaries = [
            boundary
            for boundary in boundaries
            if len(boundary.production_dependents)
            <= args.max_production_dependents
        ]
    write_tsv(boundaries)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: maintain/src/report/crate_boundaries.rs
use std::collections::{BTreeMap, BTreeSet};
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;
use std::sync::LazyLock;

use regex::Regex;
use serde::Deserialize;

use crate::MaintainResult;

static PUBLIC_ITEM: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(
        r"^pub\s+(?:async\s+|unsafe\s+|const\s+|extern\s+)*(?:struct|enum|union|trait|type|const|static|fn|mod|use)\b",
    )
    .expect("valid public item regex")
});

#[derive(Debug, Clone, Default)]
pub struct Options {
    pub max_rust_loc: Option<usize>,
    pub max_production_dependents: Option<usize>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CrateBoundary {
    pub name: String,
    pub rust_loc: usize,
    pub public_items: usize,
    pub production_dependencies: Vec<String>,
    pub production_dependents: Vec<String>,
    pub dev_only_dependents: Vec<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct DependencyKind {
    #[serde(default)]
    pub kind: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CargoDependency {
    pub name: String,
    #[serde(default)]
    pub kind: Option<String>,
    #[serde(default)]
    pub dep_kinds: Vec<DependencyKind>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CargoPackage {
    pub name: String,
    pub id: String,
    pub manifest_path: PathBuf,
    #[serde(default)]
    pub dependencies: Vec<CargoDependency>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CargoMetadata {
    pub packages: Vec<CargoPackage>,
    pub workspace_members: Vec<String>,
}

fn cargo_metadata(root: &Path) -> MaintainResult<CargoMetadata> {
    let output = Command::new("cargo")
        .args(["metadata", "--no-deps", "--format-version", "1"])
        .current_dir(root)
        .output()
        .map_err(|error| format!("failed to run cargo metadata: {error}"))?;
    if !output.status.success() {
        return Err(format!(
            "cargo metadata failed with {}:\n{}",
            output.status,
            String::from_utf8_lossy(&output.stderr)
        ));
    }
    serde_json::from_slice(&output.stdout)
        .map_err(|error| format!("failed to decode cargo metadata: {error}"))
}

fn dependency_kinds(dependency: &CargoDependency) -> BTreeSet<String> {
    if let Some(kind) = &dependency.kind {
        return BTreeSet::from([kind.clone()]);
    }
    if !dependency.dep_kinds.is_empty() {
        return dependency
            .dep_kinds
            .iter()
            .map(|entry| entry.kind.clone().unwrap_or_else(|| "normal".to_owned()))
            .collect();
    }
    BTreeSet::from(["normal".to_owned()])
}

fn collect_rust_sources(root: &Path, paths: &mut Vec<PathBuf>) -> MaintainResult<()> {
    for entry in
        fs::read_dir(root).map_err(|error| format!("failed to read {}: {error}", root.display()))?
    {
        let entry =
            entry.map_err(|error| format!("failed to inspect {}: {error}", root.display()))?;
        let file_type = entry
            .file_type()
            .map_err(|error| format!("failed to inspect {}: {error}", entry.path().display()))?;
        if file_type.is_dir() {
            collect_rust_sources(&entry.path(), paths)?;
        } else if file_type.is_file()
            && entry
                .path()
                .extension()
                .is_some_and(|extension| extension == "rs")
        {
            paths.push(entry.path());
        }
    }
    Ok(())
}

pub fn rust_source_metrics(crate_root: &Path) -> MaintainResult<(usize, usize)> {
    let source_root = crate_root.join("src");
    if !source_root.is_dir() {
        return Ok((0, 0));
    }
    let mut paths = Vec::new();
    collect_rust_sources(&source_root, &mut paths)?;
    paths.sort();
    let mut rust_loc = 0;
    let mut public_items = 0;
    for path in paths {
        let source = fs::read_to_string(&path)
            .map_err(|error| format!("failed to read {}: {error}", path.display()))?;
        for line in source
            .lines()
            .map(str::trim)
            .filter(|line| !line.is_empty())
        {
            rust_loc += 1;
            if PUBLIC_ITEM.is_match(line) {
                public_items += 1;
            }
        }
    }
    Ok((rust_loc, public_items))
}

pub fn workspace_boundaries(metadata: &CargoMetadata) -> MaintainResult<Vec<CrateBoundary>> {
    let members = metadata
        .workspace_members
        .iter()
        .cloned()
        .collect::<BTreeSet<_>>();
    let packages = metadata
        .packages
        .iter()
        .filter(|package| members.contains(&package.id))
        .map(|package| (package.name.clone(), package))
        .collect::<BTreeMap<_, _>>();
    let mut production_dependencies = packages
        .keys()
        .map(|name| (name.clone(), BTreeSet::new()))
        .collect::<BTreeMap<_, _>>();
    let mut production_dependents = production_dependencies.clone();
    let mut dev_dependents = production_dependencies.clone();

    for (consumer, package) in &packages {
        for dependency in &package.dependencies {
            let provider = &dependency.name;
            if !packages.contains_key(provider) {
                continue;
            }
            let kinds = dependency_kinds(dependency);
            if kinds.iter().any(|kind| kind != "dev") {
                production_dependencies
                    .get_mut(consumer)
                    .expect("workspace consumer set")
                    .insert(provider.clone());
                production_dependents
                    .get_mut(provider)
                    .expect("workspace provider set")
                    .insert(consumer.clone());
            }
            if kinds.contains("dev") {
                dev_dependents
                    .get_mut(provider)
                    .expect("workspace provider dev set")
                    .insert(consumer.clone());
            }
        }
    }

    let mut boundaries = Vec::new();
    for (name, package) in packages {
        let crate_root = package.manifest_path.parent().ok_or_else(|| {
            format!(
                "manifest has no parent: {}",
                package.manifest_path.display()
            )
        })?;
        let (rust_loc, public_items) = rust_source_metrics(crate_root)?;
        let production = &production_dependents[&name];
        let dev_only = dev_dependents[&name]
            .difference(production)
            .cloned()
            .collect();
        boundaries.push(CrateBoundary {
            name: name.clone(),
            rust_loc,
            public_items,
            production_dependencies: production_dependencies[&name].iter().cloned().collect(),
            production_dependents: production.iter().cloned().collect(),
            dev_only_dependents: dev_only,
        });
    }
    Ok(boundaries)
}

fn joined(values: &[String]) -> String {
    if values.is_empty() {
        "-".to_owned()
    } else {
        values.join(",")
    }
}

pub fn tsv(boundaries: &[CrateBoundary]) -> String {
    let mut output = String::from(
        "crate\trust_loc\tpublic_items\tproduction_dependencies\tproduction_dependents\tdev_only_dependents\n",
    );
    for boundary in boundaries {
        output.push_str(&format!(
            "{}\t{}\t{}\t{}\t{}\t{}\n",
            boundary.name,
            boundary.rust_loc,
            boundary.public_items,
            joined(&boundary.production_dependencies),
            joined(&boundary.production_dependents),
            joined(&boundary.dev_only_dependents),
        ));
    }
    output
}

pub fn run(root: &Path, options: &Options) -> MaintainResult<()> {
    let mut boundaries = workspace_boundaries(&cargo_metadata(root)?)?;
    boundaries.retain(|boundary| {
        options
            .max_rust_loc
            .is_none_or(|maximum| boundary.rust_loc <= maximum)
            && options
                .max_production_dependents
                .is_none_or(|maximum| boundary.production_dependents.len() <= maximum)
    });
    print!("{}", tsv(&boundaries));
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_support::TestDirectory;

    #[test]
    fn counts_non_empty_source_lines_and_public_items() {
        let directory = TestDirectory::new("source-metrics");
        directory.write(
            "src/lib.rs",
            "pub struct Public;\n\nstruct Private;\npub(crate) fn helper() {}\n",
        );
        assert_eq!(
            rust_source_metrics(directory.path()).expect("measure source"),
            (3, 1)
        );
    }

    #[test]
    fn separates_production_and_dev_only_dependents() {
        let directory = TestDirectory::new("crate-boundaries");
        let package = |name: &str, dependencies: Vec<CargoDependency>| CargoPackage {
            name: name.to_owned(),
            id: name.to_owned(),
            manifest_path: directory.path().join(name).join("Cargo.toml"),
            dependencies,
        };
        for name in ["provider", "consumer", "dev-consumer"] {
            directory.write(&format!("{name}/src/lib.rs"), "");
        }
        let dependency = |kind: Option<&str>| CargoDependency {
            name: "provider".to_owned(),
            kind: kind.map(str::to_owned),
            dep_kinds: Vec::new(),
        };
        let metadata = CargoMetadata {
            packages: vec![
                package("provider", Vec::new()),
                package("consumer", vec![dependency(None), dependency(Some("dev"))]),
                package("dev-consumer", vec![dependency(Some("dev"))]),
            ],
            workspace_members: vec![
                "provider".to_owned(),
                "consumer".to_owned(),
                "dev-consumer".to_owned(),
            ],
        };
        let boundaries = workspace_boundaries(&metadata)
            .expect("compute boundaries")
            .into_iter()
            .map(|boundary| (boundary.name.clone(), boundary))
            .collect::<BTreeMap<_, _>>();

        assert_eq!(boundaries["provider"].production_dependents, ["consumer"]);
        assert_eq!(boundaries["provider"].dev_only_dependents, ["dev-consumer"]);
        assert_eq!(boundaries["consumer"].production_dependencies, ["provider"]);
    }
}
